### PR TITLE
Multipart copy method

### PR DESCRIPTION
**Pull request by nmuntyanov**
fix https://github.com/async-aws/aws/issues/1589

### CODE REVIEW

File: src/Integration/Aws/SimpleS3/src/SimpleS3Client.php [hunk 1]
```diff
@@ -47,6 +52,71 @@ public function has(string $bucket, string $key): bool
...
+                array_merge($options, ['Bucket' => $destBucket, 'Key' => $destKey])
+            )
+        )->getUploadId();
+
+        $bytePosition = 0;
+        $parts = [];
+        for ($i = 1; $bytePosition < $contentLength; ++$i) {
+            $startByte = $bytePosition;
+            $endByte = $bytePosition + $partSize - 1 >= $contentLength ? $contentLength - 1 : $bytePosition + $partSize - 1;
+            $parts[] = $this->doMultipartCopy($destBucket, $destKey, $uploadId, $i, sprintf('%s/%s', $srcBucket, $srcKey), $startByte, $endByte);
```
Comment (jderusse) [inline src/Integration/Aws/SimpleS3/src/SimpleS3Client.php:107]: would it make sens to run this in parallel?

File: src/Integration/Aws/SimpleS3/src/SimpleS3Client.php [hunk 2]
```diff
@@ -195,4 +265,28 @@ private function doSmallFileUpload(array $options, string $bucket, string $key,
             'Body' => $object,
         ]));
     }
+
+    private function doMultipartCopy(string $bucket, string $key, string $uploadId, int $partNumber, string $copySource, int $startByte, int $endByte): CompletedPart
+    {
+        try {
+            $response = $this->uploadPartCopy(
```
Comment (jderusse) [inline src/Integration/Aws/SimpleS3/src/SimpleS3Client.php:272]: this operation requires the (not yet release) next version of S3.
You need to either upgrade composer.json `"async-aws/s3": "^2.1"`

**Review by jderusse — COMMENTED** [review]
You should split this PR in 2:
- adding the `uploadPartCopy` operation in S3
- adding a new method in SimpleS3 that use the new method of S3

File: src/Integration/Aws/SimpleS3/src/SimpleS3Client.php (hunk 1 above)
Comment (nmuntyanov) [reply src/Integration/Aws/SimpleS3/src/SimpleS3Client.php:107]: makes sense
but i have no idea how to implement this))

File: src/Integration/Aws/SimpleS3/src/SimpleS3Client.php (hunk 2 above)
Comment (nmuntyanov) [reply src/Integration/Aws/SimpleS3/src/SimpleS3Client.php:272]: https://github.com/async-aws/aws/pull/1591
First one

File: src/Integration/Aws/SimpleS3/src/SimpleS3Client.php (hunk 1 above)
Comment (jderusse) [reply src/Integration/Aws/SimpleS3/src/SimpleS3Client.php:107]: Avoid consuming the response: When you access a property (calling a getter) the response blocks until the response is fully processed.

once again, I don't know if it's better to run theses requests in sequence or parallel.
(Maybe run too many requests in parallel have worse performance) => you need to check AWS recommendations for this.


to process requests in parallel you should do something like
```
for (...) {
  $responses[] = $client->uploadPartCopy(...)
}

$success = true;
foreach ($responses as $response) {
  try {
    $copyPartResult = $response->getCopyPartResult();
    $parts[] = new CompletedPart(['ETag' => $copyPartResult->getEtag(), 'PartNumber' => $partNumber]);
  } catch (\Throwable $e) {
    $success = false;
    break'
  }
}

if (!$success) {
  $this->abortMultipartUpload(['Bucket' => $bucket, 'Key' => $key, 'UploadId' => $uploadId]);
  foreach ($responses as $response) {
    try {
      $response->cancel();
    } catch (\Throwable $e) {
      // ...
    }
  }

   throw ...;
}
```

Comment (nmuntyanov) [reply src/Integration/Aws/SimpleS3/src/SimpleS3Client.php:107]: Got it!
Seems like the only limit is 10 000 connections (equal to parts)
https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html
Anyway, think that 10k connections, even without body, it is too much.

AWS SDK has concurrency option for such copy
Will see how to implement it here

**Pull request closed without merging**